File: frontend/lib/api/contracts.ts
/**
 * Contract Address API Client
 *
 * Fetches contract addresses from the backend PostgreSQL database via API.
 *
 * IMPORTANT: NO HARDCODED FALLBACK ADDRESSES
 * All contract addresses MUST come from the database. This ensures:
 * - Single source of truth for contract addresses
 * - OCP (Open-Closed Principle) compliance - new deployments only require DB updates
 * - No risk of stale hardcoded addresses causing bugs
 *
 * If the API is unavailable, the error should propagate and be handled by the UI
 * with appropriate error states rather than silently using potentially outdated addresses.
 */

import { type Address } from 'viem';

const API_URL = process.env.NEXT_PUBLIC_API_URL || 'http://localhost:8080';

// ============================================================================
// Response Types from API
// ============================================================================

export type NetworkConfigResponse = {
  id: string;
  chain_id: number;
  network_name: string;
  display_name: string;
  rpc_url: string | null;
  explorer_url: string | null;
  default_deployer: string | null;
  is_testnet: boolean;
  is_active: boolean;
  created_at: string;
  updated_at: string;
};

export type ContractMappingResponse = {
  id: string;
  solidity_name: string;
  db_name: string;
  display_name: string;
  category: string;
  description: string | null;
  is_required: boolean;
  sort_order: number;
  created_at: string;
};

export type ContractAddressResponse = {
  id: string;
  chain_id: number;
  contract_mapping_id: string;
  db_name: string;
  solidity_name: string;
  address: string;
  deployment_tx_hash: string | null;
  deployment_block: number | null;
  abi_version: string;
  status: string;
  is_primary: boolean;
  deployed_by: string | null;
  notes: string | null;
  created_at: string;
  updated_at: string;
};

export type ContractHistoryResponse = {
  id: string;
  contract_id: string;
  old_address: string | null;
  new_address: string;
  change_reason: string | null;
  changed_by: string;
  changed_at: string;
};

export type DeploymentConfig = {
  network: NetworkConfigResponse;
  mappings: ContractMappingResponse[];
  contracts: ContractAddressResponse[];
};

// Dynamic contract addresses - keys come from database
export type ContractAddresses = Record<string, Address>;

// ============================================================================
// API Response Wrapper
// ============================================================================

type ApiResponse<T> = {
  success: boolean;
  data?: T;
  message?: string;
  error?: string;
};

// ============================================================================
// API Functions
// ============================================================================

/**
 * Fetch all active networks
 */
export async function fetchNetworks(): Promise<NetworkConfigResponse[]> {
  const response = await fetch(`${API_URL}/api/v1/networks`);
  if (!response.ok) {
    throw new Error(`Failed to fetch networks: ${response.statusText}`);
  }
  const data: ApiResponse<{ networks: NetworkConfigResponse[]; total: number }> =
    await response.json();
  if (!data.success) {
    throw new Error(data.error || 'Failed to fetch networks');
  }
  return data.data?.networks || [];
}

/**
 * Fetch network config by chain ID
 */
export async function fetchNetwork(chainId: number): Promise<NetworkConfigResponse> {
  const response = await fetch(`${API_URL}/api/v1/networks/${chainId}`);
  if (!response.ok) {
    if (response.status === 404) {
      throw new Error(`Network not found for chain ID: ${chainId}`);
    }
    throw new Error(`Failed to fetch network: ${response.statusText}`);
  }
  const data: ApiResponse<NetworkConfigResponse> = await response.json();
  if (!data.success) {
    throw new Error(data.error || 'Failed to fetch network');
  }
  return data.data!;
}

/**
 * Fetch all contract name mappings
 */
export async function fetchContractMappings(): Promise<ContractMappingResponse[]> {
  const response = await fetch(`${API_URL}/api/v1/contracts/mappings`);
  if (!response.ok) {
    throw new Error(`Failed to fetch mappings: ${response.statusText}`);
  }
  const data: ApiResponse<{ mappings: ContractMappingResponse[]; total: number }> =
    await response.json();
  if (!data.success) {
    throw new Error(data.error || 'Failed to fetch mappings');
  }
  return data.data?.mappings || [];
}

/**
 * Fetch full deployment config (network + mappings + contracts)
 */
export async function fetchDeploymentConfig(chainId: number): Promise<DeploymentConfig> {
  const response = await fetch(`${API_URL}/api/v1/contracts/config/${chainId}`);
  if (!response.ok) {
    if (response.status === 404) {
      throw new Error(`Network not found for chain ID: ${chainId}`);
    }
    throw new Error(`Failed to fetch config: ${response.statusText}`);
  }
  const data: ApiResponse<DeploymentConfig> = await response.json();
  if (!data.success) {
    throw new Error(data.error || 'Failed to fetch config');
  }
  return data.data!;
}

/**
 * Fetch all contracts for a chain
 */
export async function fetchContractsList(
  chainId: number
): Promise<ContractAddressResponse[]> {
  const response = await fetch(`${API_URL}/api/v1/contracts/${chainId}`);
  if (!response.ok) {
    throw new Error(`Failed to fetch contracts: ${response.statusText}`);
  }
  const data: ApiResponse<{ contracts: ContractAddressResponse[]; total: number }> =
    await response.json();
  if (!data.success) {
    throw new Error(data.error || 'Failed to fetch contracts');
  }
  return data.data?.contracts || [];
}

/**
 * Fetch contract addresses from the PostgreSQL database via API.
 * Returns a dynamic Record<string, Address> built from database data.
 *
 * NO FALLBACKS - All addresses come from the database.
 * If the API fails, the error propagates to be handled by the UI.
 */
export async function fetchContractAddresses(chainId: number): Promise<ContractAddresses> {
  const contracts = await fetchContractsList(chainId);

  // Build dynamic object from API response
  return contracts.reduce(
    (acc: ContractAddresses, c: ContractAddressResponse) => ({
      ...acc,
      [c.db_name]: c.address as Address,
    }),
    {} as ContractAddresses
  );
}

/**
 * Fetch a specific contract by chain ID and DB name
 */
export async function fetchContract(
  chainId: number,
  dbName: string
): Promise<ContractAddressResponse> {
  const response = await fetch(`${API_URL}/api/v1/contracts/${chainId}/${dbName}`);
  if (!response.ok) {
    if (response.status === 404) {
      throw new Error(`Contract not found: ${dbName} on chain ${chainId}`);
    }
    throw new Error(`Failed to fetch contract: ${response.statusText}`);
  }
  const data: ApiResponse<ContractAddressResponse> = await response.json();
  if (!data.success) {
    throw new Error(data.error || 'Failed to fetch contract');
  }
  return data.data!;
}

/**
 * Fetch contract deployment history
 */
export async function fetchContractHistory(
  contractId: string,
  limit: number = 20
): Promise<ContractHistoryResponse[]> {
  const response = await fetch(
    `${API_URL}/api/v1/contracts/history/${contractId}?limit=${limit}`
  );
  if (!response.ok) {
    if (response.status === 404) {
      throw new Error(`Contract not found: ${contractId}`);
    }
    throw new Error(`Failed to fetch history: ${response.statusText}`);
  }
  const data: ApiResponse<{
    contract_id: string;
    contract_name: string;
    chain_id: number;
    history: ContractHistoryResponse[];
    total: number;
  }> = await response.json();
  if (!data.success) {
    throw new Error(data.error || 'Failed to fetch history');
  }
  return data.data?.history || [];
}

File: backend/internal/repository/contracts.go
// Package repository defines the interfaces for data access
package repository

import (
	"context"
	"time"
)

// ContractRepository defines the contract for contract address data operations
type ContractRepository interface {
	// Network configuration
	GetNetworkByChainID(ctx context.Context, chainID int64) (*NetworkConfig, error)
	GetNetworkByName(ctx context.Context, name string) (*NetworkConfig, error)
	GetActiveNetworks(ctx context.Context) ([]*NetworkConfig, error)

	// Contract name mappings (read from DB, no hardcoded maps)
	GetAllMappings(ctx context.Context) ([]*ContractMapping, error)
	GetMappingBySolidityName(ctx context.Context, name string) (*ContractMapping, error)
	GetMappingByDBName(ctx context.Context, dbName string) (*ContractMapping, error)

	// Contract addresses
	GetByChainID(ctx context.Context, chainID int64) ([]*ContractAddress, error)
	GetByChainAndDBName(ctx context.Context, chainID int64, dbName string) (*ContractAddress, error)
	GetByID(ctx context.Context, id string) (*ContractAddress, error)
	Upsert(ctx context.Context, contract *ContractAddressUpsert) (*ContractAddress, error)

	// Deployment history (audit trail)
	GetHistory(ctx context.Context, contractID string, limit int) ([]*ContractAddressHistory, error)

	// Combined endpoint for deploy scripts - returns everything needed for deployment
	GetDeploymentConfig(ctx context.Context, chainID int64) (*DeploymentConfig, error)
}

// NetworkConfig represents per-network configuration from DB
type NetworkConfig struct {
	ID              string    `json:"id" db:"id"`
	ChainID         int64     `json:"chain_id" db:"chain_id"`
	NetworkName     string    `json:"network_name" db:"network_name"`
	DisplayName     string    `json:"display_name" db:"display_name"`
	RPCUrl          *string   `json:"rpc_url,omitempty" db:"rpc_url"`
	ExplorerUrl     *string   `json:"explorer_url,omitempty" db:"explorer_url"`
	DefaultDeployer *string   `json:"default_deployer,omitempty" db:"default_deployer"`
	IsTestnet       bool      `json:"is_testnet" db:"is_testnet"`
	IsActive        bool      `json:"is_active" db:"is_active"`
	CreatedAt       time.Time `json:"created_at" db:"created_at"`
	UpdatedAt       time.Time `json:"updated_at" db:"updated_at"`
}

// ContractMapping represents Solidity→DB name mapping from DB
type ContractMapping struct {
	ID           string    `json:"id" db:"id"`
	SolidityName string    `json:"solidity_name" db:"solidity_name"`
	DBName       string    `json:"db_name" db:"db_name"`
	DisplayName  string    `json:"display_name" db:"display_name"`
	Category     string    `json:"category" db:"category"`
	Description  *string   `json:"description,omitempty" db:"description"`
	IsRequired   bool      `json:"is_required" db:"is_required"`
	SortOrder    int       `json:"sort_order" db:"sort_order"`
	CreatedAt    time.Time `json:"created_at" db:"created_at"`
}

// ContractAddress represents a deployed contract from DB
type ContractAddress struct {
	ID                string    `json:"id" db:"id"`
	ChainID           int64     `json:"chain_id" db:"chain_id"`
	ContractMappingID string    `json:"contract_mapping_id" db:"contract_mapping_id"`
	DBName            string    `json:"db_name" db:"db_name"`                   // Joined from contract_mappings
	SolidityName      string    `json:"solidity_name" db:"solidity_name"`       // Joined from contract_mappings
	Address           string    `json:"address" db:"address"`
	DeploymentTxHash  *string   `json:"deployment_tx_hash,omitempty" db:"deployment_tx_hash"`
	DeploymentBlock   *int64    `json:"deployment_block,omitempty" db:"deployment_block"`
	ABIVersion        string    `json:"abi_version" db:"abi_version"`
	Status            string    `json:"status" db:"status"`
	IsPrimary         bool      `json:"is_primary" db:"is_primary"`
	DeployedBy        *string   `json:"deployed_by,omitempty" db:"deployed_by"`
	Notes             *string   `json:"notes,omitempty" db:"notes"`
	CreatedAt         time.Time `json:"created_at" db:"created_at"`
	UpdatedAt         time.Time `json:"updated_at" db:"updated_at"`
}

// ContractAddressUpsert represents data for creating/updating a contract address
type ContractAddressUpsert struct {
	ChainID           int64   `json:"chain_id" binding:"required"`
	ContractMappingID string  `json:"contract_mapping_id" binding:"required"`
	Address           string  `json:"address" binding:"required"`
	DeploymentTxHash  *string `json:"deployment_tx_hash,omitempty"`
	DeploymentBlock   *int64  `json:"deployment_block,omitempty"`
	ABIVersion        *string `json:"abi_version,omitempty"`
	DeployedBy        *string `json:"deployed_by,omitempty"`
	Notes             *string `json:"notes,omitempty"`
}

// ContractAddressHistory represents an audit trail entry for address changes
type ContractAddressHistory struct {
	ID           string    `json:"id" db:"id"`
	ContractID   string    `json:"contract_id" db:"contract_id"`
	OldAddress   *string   `json:"old_address,omitempty" db:"old_address"`
	NewAddress   string    `json:"new_address" db:"new_address"`
	ChangeReason *string   `json:"change_reason,omitempty" db:"change_reason"`
	ChangedBy    string    `json:"changed_by" db:"changed_by"`
	ChangedAt    time.Time `json:"changed_at" db:"changed_at"`
}

// DeploymentConfig is the combined config returned to deploy scripts
type DeploymentConfig struct {
	Network   *NetworkConfig     `json:"network"`
	Mappings  []*ContractMapping `json:"mappings"`
	Contracts []*ContractAddress `json:"contracts"`
}
